xray/segment.go

```diff
@@ -116,7 +135,11 @@ func BeginSegmentWithSampling(ctx context.Context, name string, r *http.Request,
 		}()
 	}
 
-	return context.WithValue(ctx, ContextKey, seg), seg
+	if seg.Sampled {
+		return context.WithValue(ctx, ContextKey, seg), seg
+	}
+
+	return BeginDummySegment(ctx, "DummySegment")
```

Comment:
If the Segment is dummy, adding annotation and metadata should be no-op to save the memory footprint.